### DISCUSSION TITLE
Inline and static sampler states.

### DISCUSSION THREAD

**Discussion by Gavin-Williams** [amerkoleci/Vortice.Windows#421] [General]
Hi,

If I want to quickly create a sampler state in shader, how do I do that? I tried this...

```
SamplerState TextureSampler
{
    Filter = MIN_MAG_MIP_ANISOTROPIC;
    AddressU = Wrap;
    AddressV = Wrap;
};
...
float3 diffuse = textures.Sample(TextureSampler, float3(input.tex, 0));
```
But that doesn't do anything. I also tried making it static, and that didn't work either. Could someone guide me as to how I get this to work (DirectX 11) and an explanation of how to use static objects in hlsl as well?

I know how to upload a render-state, but I'm sure I've done it this way before, and looking online I see ppl talking about it, but I can't get an actual answer to how to create a sampler state in hlsl.

**Comment by Gavin-Williams** [comment]
I thought maybe something like this could work...

```
SamplerState AnisotropicSampler
{
    Filter = MIN_MAG_MIP_ANISOTROPIC;
    AddressU = Wrap;
    AddressV = Wrap;
};
...
sampler TextureSampler : register(s0) = AnisotropicSampler;
```
But I get an error, initial value must be a literal expression. It looks pretty literal to me. So I guess I don't understand this aspect of the language at all. I guess I've created a variable 'AnisostropicSampler'. Though I also tried assigning the curley brace code directly to TextureSampler and that didn't work.

**Reply by Gavin-Williams** [reply]
Reading this... https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-state-object
> 
With shader models 6.3 and later, applications have the convenience and flexibility of being able to define DXR state objects directly in HLSL shader code in addition to using Direct3D 12 APIs.
Makes me think that it might be a Dx12 feature only. I'm using shader model 5. If this is the same feature - then that might be answer to why it's not woking.

**Reply by Gavin-Williams** [reply]
Looking at this page, it says that I can do exactly what I did in the Original Post, in DirectX 10. So why isn't this working?

https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-sampler

**Reply by amerkoleci** [reply]
Hi,
I know that the legacy effect framework had inline support for samplers and other states (rasterizer, depth stencil state, blend), this is not supported under normal shader compilation.

D3d12 and shader model 6.0+ allows you to define root signature in shader, thus static samples.

So I would suggest other way to handle this or implement own parser logic. 

Hope I helped out.
